tests/mir_test_framework/window_management_test_harness.cpp

```diff
@@ -57,90 +59,304 @@ class CachedSurfaceData
+    {
+        tools.modify_window(window_info, modifications);
+        on_change();
+    }
+
+    void place_and_size_for_state(miral::WindowSpecification& modifications, miral::WindowInfo const& window_info) const override
+    {
+        tools.place_and_size_for_state(modifications, window_info);
+        on_change();
+    }
+
+    auto create_workspace() -> std::shared_ptr<miral::Workspace> override
+    {
+        auto workspace = tools.create_workspace();
+        on_change();
+        return workspace;
+    }
+
+    void add_tree_to_workspace(miral::Window const& window, std::shared_ptr<miral::Workspace> const& workspace) override
+    {
+        tools.add_tree_to_workspace(window, workspace);
+        on_change();
+    }
+
+    void remove_tree_from_workspace(miral::Window const& window, std::shared_ptr<miral::Workspace> const& workspace) override
+    {
+        tools.remove_tree_from_workspace(window, workspace);
+        on_change();
+    }
+
+    void move_workspace_content_to_workspace(
+        std::shared_ptr<miral::Workspace> const& to_workspace,
+        std::shared_ptr<miral::Workspace> const& from_workspace) override
+    {
+        tools.move_workspace_content_to_workspace(to_workspace, from_workspace);
+        on_change();
+    }
+
+    void for_each_workspace_containing(
+        miral::Window const& window,
+        std::function<void(std::shared_ptr<miral::Workspace> const&)> const& callback) override
+    {
+        tools.for_each_workspace_containing(window, callback);
+    }
+
+    void for_each_window_in_workspace(
+        std::shared_ptr<miral::Workspace> const& workspace,
+        std::function<void(miral::Window const& window)> const& callback) override
+    {
+        tools.for_each_window_in_workspace(workspace, callback);
+    }
+
+    void invoke_under_lock(std::function<void()> const& callback) override
+    {
+        tools.invoke_under_lock(callback);
+    }
+
+    void move_cursor_to(mir::geometry::PointF point) override
+    {
+        tools.move_cursor_to(point);
+        on_change();
+    }
+
+    void drag_window(
+        miral::Window const& window, mir::geometry::Displacement& movement) override
+    {
+        tools.drag_window(window, movement);
+        on_change();
+    }
+
+    auto active_application_zone() -> miral::Zone override
+    {
+        return tools.active_application_zone();
+    }
+private:
+    miral::WindowManagerTools tools;
+    std::function<void()> on_change;
+};
+}
+
+class mir_test_framework::WindowManagementTestHarness::Self
+{
+public:
+    explicit Self(mir::Server& server)
+        : server(server),
+          surface_observer(std::make_shared<SurfaceCloseRequestedObserver>([this](ms::Surface const* surf)
+          {
+              on_surface_closed(surf);
+          }))
+    {
+    }
+
+    void process_pending()
+    {
+        for (auto const& action : pending_actions)
+            action();
+        pending_actions.clear();
```

Comment:
I can see that we currently won't update `pending_actions` while iterating, but it seems plausible that future tests will lead `pending_actions` that lead recursively to more `pending_actions`.
```suggestion
        decltype(pending_actions) actions;
        actions.swap(pending_actions);
        for (auto const& action : actions)
            action();
```

Reply:
Yup that makes sense